app/services/RestaurantService.java

```diff
@@ -30,228 +30,288 @@
+
+    @Inject
+    private RestaurantService() {
+    }
+
+    /**
+     * Create restaurant boolean.
+     *
+     * @param restaurant the restaurant
+     * @throws Exception the exception
+     */
+    public Boolean createRestaurant(final Restaurant restaurant) throws Exception {
+        getSession().save(restaurant);
+        return true;
+    }
+
+    /**
+     * Edit restaurant boolean.
+     *
+     * @param restaurant the restaurant
+     * @throws Exception the exception
+     */
+    public Boolean editRestaurant(final Restaurant restaurant) throws Exception {
+        getSession().merge(restaurant);
+        return true;
+    }
+
+    /**
+     * Delete restaurant boolean.
+     *
+     * @param id the id
+     * @throws Exception the exception
+     */
+    public Boolean deleteRestaurant(final UUID id) throws Exception {
+        Restaurant restaurant = (Restaurant) getSession().createCriteria(Restaurant.class)
+                .add(Restrictions.eq("id", id))
+                .uniqueResult();
+
+        getSession().delete(restaurant);
+        return true;
+    }
+
+    /**
+     * Find restaurants with filter pagination adapter.
+     *
+     * @param restaurantFilter the restaurant filter
+     * @return the pagination adapter
+     */
+    @SuppressWarnings("unchecked")
+    public PaginationAdapter<Restaurant> findRestaurantsWithFilter(final RestaurantFilter restaurantFilter) {
+        Criteria criteria = getSession().createCriteria(Restaurant.class);
+
+        if (restaurantFilter.name != null) {
+            criteria.add(Restrictions.ilike("name", restaurantFilter.name, MatchMode.ANYWHERE));
+        }
+
+        if (restaurantFilter.cityId != null) {
+            criteria.add(Restrictions.eq("city.id", restaurantFilter.cityId));
+        }
+
+        if (restaurantFilter.price != 0) {
+            criteria.add(Restrictions.eq("priceRange", restaurantFilter.price));
+        }
+
+        if (restaurantFilter.cuisine != null && !restaurantFilter.cuisine.equalsIgnoreCase("")) {
+            Criteria cuisine = criteria.createCriteria("cuisines");
+            String[] cuisines = restaurantFilter.cuisine.split(",");
+            for (String singleCuisine : cuisines) {
+                cuisine.add(Restrictions.eq("name", singleCuisine));
+            }
+        }
+
+        Long numberOfPages = ((Long) criteria.setProjection(Projections.rowCount()).uniqueResult()) / restaurantFilter.pageSize;
+
+        criteria.setProjection(null)
+                .setFirstResult((restaurantFilter.pageNumber - 1) * restaurantFilter.pageSize)
+                .setMaxResults(restaurantFilter.pageSize);
+
+        if (restaurantFilter.sortBy.equals("price")) {
+            criteria.addOrder(Order.desc("priceRange"));
+        }
+
+        criteria.addOrder(Order.asc("name"));
+
+        List<Restaurant> restaurants = criteria.list();
+
+        if (restaurantFilter.rating != null && restaurantFilter.rating != 0) {
```

Comment:
How about get all reviews that meet filter conditions, and from reviews list get list of restaurant ids that should be returned?